File: backend/sombra/shadows.py
"""Shadow engine: sun position, shadow polygons, per-edge sun exposure.

All geometric functions operate in a projected CRS (meters). The caller is
responsible for projecting geometries before calling in here.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime

from astral import Observer
from astral.sun import azimuth as astral_azimuth
from astral.sun import elevation as astral_elevation
from shapely.geometry import LineString, MultiPolygon, Point, Polygon
from shapely.geometry.base import BaseGeometry
from shapely.affinity import translate
from shapely.ops import unary_union
from shapely.strtree import STRtree

from . import config


@dataclass(frozen=True)
class SunPosition:
    elevation_deg: float
    azimuth_deg: float

    @property
    def is_low(self) -> bool:
        """Sun low enough that everything is effectively shaded (or it's night)."""
        return self.elevation_deg <= config.LOW_SUN_ELEVATION_DEG


def sun_position(lat: float, lon: float, when: datetime) -> SunPosition:
    """Sun elevation/azimuth for a tz-aware datetime at (lat, lon)."""
    obs = Observer(latitude=lat, longitude=lon)
    return SunPosition(
        elevation_deg=astral_elevation(obs, when),
        azimuth_deg=astral_azimuth(obs, when),
    )


def parse_height(tags: dict) -> float:
    """Building height in meters: height -> building:levels * 3.2 -> default."""
    raw = tags.get("height")
    if raw is not None:
        try:
            return float(str(raw).lower().replace("m", "").replace(",", ".").strip())
        except ValueError:
            pass
    raw = tags.get("building:levels")
    if raw is not None:
        try:
            levels = float(str(raw).split(";")[0].replace(",", "."))
            if levels > 0:
                return levels * config.METERS_PER_LEVEL
        except ValueError:
            pass
    return config.DEFAULT_BUILDING_HEIGHT_M


def shadow_offset(height_m: float, sun: SunPosition) -> tuple[float, float]:
    """(dx, dy) in meters from an object's base to the tip of its shadow.

    The shadow points away from the sun: direction = azimuth + 180deg.
    dx is easting, dy is northing (projected CRS).
    """
    if sun.is_low:
        return (0.0, 0.0)
    length = min(height_m / math.tan(math.radians(sun.elevation_deg)), config.MAX_SHADOW_LEN_M)
    direction = math.radians(sun.azimuth_deg + 180.0)
    return (length * math.sin(direction), length * math.cos(direction))


def _hull_shadow(geom: BaseGeometry, dx: float, dy: float) -> BaseGeometry:
    """Shadow of a convex-ish footprint: hull of footprint + translated footprint.

    NOTE: over-approximates shadows of strongly concave buildings (fills inner
    courtyards). Acceptable for MVP -- it errs toward predicting *more* shade
    near building edges where real shade usually exists.
    """
    moved = translate(geom, xoff=dx, yoff=dy)
    return unary_union([geom, moved]).convex_hull


def building_shadow(footprint: BaseGeometry, height_m: float, sun: SunPosition) -> BaseGeometry | None:
    if sun.is_low or footprint.is_empty:
        return None
    dx, dy = shadow_offset(height_m, sun)
    parts = footprint.geoms if isinstance(footprint, MultiPolygon) else [footprint]
    shadows = [_hull_shadow(p, dx, dy) for p in parts if not p.is_empty]
    if not shadows:
        return None
    return unary_union(shadows)


def tree_shadow(geom: BaseGeometry, sun: SunPosition) -> BaseGeometry | None:
    """Shadow for natural=tree (point) or natural=tree_row (line)."""
    if sun.is_low or geom is None or geom.is_empty:
        return None
    dx, dy = shadow_offset(config.TREE_HEIGHT_M, sun)
    if isinstance(geom, Point):
        crown = geom.buffer(config.TREE_RADIUS_M, quad_segs=4)
    elif isinstance(geom, LineString):
        crown = geom.buffer(config.TREE_ROW_BUFFER_M, quad_segs=4)
    else:  # polygons (rare for trees) -- treat like a crown already
        crown = geom
    return _hull_shadow(crown, dx, dy)


class ShadeIndex:
    """Spatial index over shadow polygons for fast per-edge queries."""

    def __init__(self, shadow_geoms: list[BaseGeometry]):
        self._geoms = [g for g in shadow_geoms if g is not None and not g.is_empty]
        self._tree = STRtree(self._geoms) if self._geoms else None

    @property
    def count(self) -> int:
        return len(self._geoms)

    def sun_fraction(self, edge_geom: LineString) -> float:
        """Fraction of the edge length exposed to the sun (0 = fully shaded)."""
        if self._tree is None:
            return 1.0
        if edge_geom.length < 1.0:
            # Degenerate edge: sample the midpoint.
            mid = edge_geom.interpolate(0.5, normalized=True)
            idx = self._tree.query(mid)
            for i in idx:
                if self._geoms[i].covers(mid):
                    return 0.0
            return 1.0
        idx = self._tree.query(edge_geom)
        if len(idx) == 0:
            return 1.0
        shade = unary_union([self._geoms[i] for i in idx])
        shaded_len = edge_geom.intersection(shade).length
        return max(0.0, min(1.0, 1.0 - shaded_len / edge_geom.length))


def build_shade_index(
    building_records: list[tuple[BaseGeometry, dict]],
    tree_geoms: list[BaseGeometry],
    sun: SunPosition,
) -> ShadeIndex:
    """building_records: (projected footprint, tags) pairs. tree_geoms: projected."""
    shadows: list[BaseGeometry] = []
    for footprint, tags in building_records:
        s = building_shadow(footprint, parse_height(tags), sun)
        if s is not None:
            shadows.append(s)
    for g in tree_geoms:
        s = tree_shadow(g, sun)
        if s is not None:
            shadows.append(s)
    return ShadeIndex(shadows)

File: web/js/shadows.js
// Shadow engine: sun position (suncalc), shadow polygons, sampled shade tests.
(function () {
  const C = SOMBRA.config, G = SOMBRA.geo;
  const S = {};

  S.sunPosition = function (lat, lon, date) {
    const p = SunCalc.getPosition(date, lat, lon);
    // suncalc azimuth: 0 = south, positive westward. Convert to compass (0 = north, clockwise).
    const azimuthDeg = ((p.azimuth * 180) / Math.PI + 180) % 360;
    const elevationDeg = (p.altitude * 180) / Math.PI;
    return { elevationDeg, azimuthDeg, isLow: elevationDeg <= C.LOW_SUN_DEG };
  };

  S.parseHeight = function (tags) {
    if (!tags) return C.DEFAULT_BUILDING_H;
    if (tags.height) {
      const h = parseFloat(String(tags.height).replace(",", ".").replace(/m/i, "").trim());
      if (isFinite(h) && h > 0) return h;
    }
    const lv = tags["building:levels"];
    if (lv) {
      const n = parseFloat(String(lv).split(";")[0].replace(",", "."));
      if (isFinite(n) && n > 0) return n * C.METERS_PER_LEVEL;
    }
    return C.DEFAULT_BUILDING_H;
  };

  S.shadowOffset = function (heightM, sun) {
    if (sun.isLow) return [0, 0];
    const L = Math.min(heightM / Math.tan((sun.elevationDeg * Math.PI) / 180), C.MAX_SHADOW_LEN_M);
    const dir = ((sun.azimuthDeg + 180) * Math.PI) / 180;
    return [L * Math.sin(dir), L * Math.cos(dir)]; // [dx east, dy north]
  };

  // casters: [{ring: [[x,y],...], h: meters}]  ->  ShadeIndex
  S.buildIndex = function (casters, sun) {
    const polys = [];
    if (!sun.isLow) {
      for (const c of casters) {
        const [dx, dy] = S.shadowOffset(c.h, sun);
        const moved = c.ring.map(([x, y]) => [x + dx, y + dy]);
        const hull = G.convexHull(c.ring.concat(moved));
        if (hull.length >= 3) polys.push({ ring: hull, bbox: G.bboxOfRing(hull) });
      }
    }
    // spatial grid buckets
    const cell = C.GRID_CELL_M, grid = new Map();
    polys.forEach((p, i) => {
      const [minx, miny, maxx, maxy] = p.bbox;
      for (let gx = Math.floor(minx / cell); gx <= Math.floor(maxx / cell); gx++)
        for (let gy = Math.floor(miny / cell); gy <= Math.floor(maxy / cell); gy++) {
          const k = gx + ":" + gy;
          if (!grid.has(k)) grid.set(k, []);
          grid.get(k).push(i);
        }
    });
    return { polys, grid, cell, count: polys.length, isLow: sun.isLow };
  };

  S.pointShaded = function (index, x, y) {
    if (index.isLow) return true;
    const k = Math.floor(x / index.cell) + ":" + Math.floor(y / index.cell);
    const bucket = index.grid.get(k);
    if (!bucket) return false;
    for (const i of bucket) {
      const p = index.polys[i];
      const b = p.bbox;
      if (x < b[0] || x > b[2] || y < b[1] || y > b[3]) continue;
      if (G.pointInPolygon(x, y, p.ring)) return true;
    }
    return false;
  };

  // Fraction of the polyline length in the SUN (1 = full sun), sampled every SAMPLE_STEP_M.
  S.sunFraction = function (index, pts) {
    if (index.isLow) return 0;
    if (index.count === 0) return 1;
    let total = 0, shaded = 0;
    for (let i = 1; i < pts.length; i++) {
      const [x1, y1] = pts[i - 1], [x2, y2] = pts[i];
      const segLen = Math.hypot(x2 - x1, y2 - y1);
      const n = Math.max(1, Math.round(segLen / C.SAMPLE_STEP_M));
      for (let s = 0; s < n; s++) {
        const t = (s + 0.5) / n;
        if (S.pointShaded(index, x1 + (x2 - x1) * t, y1 + (y2 - y1) * t)) shaded += segLen / n;
        total += segLen / n;
      }
    }
    return total === 0 ? 1 : Math.max(0, Math.min(1, 1 - shaded / total));
  };

  window.SOMBRA.shadows = S;
})();
